File: android/app/src/main/java/com/dailymemory/domain/usecase/person/DeletePersonUseCase.kt
package com.dailymemory.domain.usecase.person

import com.dailymemory.domain.repository.PersonRepository
import com.dailymemory.domain.repository.ReminderRepository
import kotlinx.coroutines.flow.first
import javax.inject.Inject

/**
 * Use case for deleting a person.
 * Also handles cleanup of related reminders.
 */
class DeletePersonUseCase @Inject constructor(
    private val personRepository: PersonRepository,
    private val reminderRepository: ReminderRepository
) {
    suspend operator fun invoke(personId: String): Result<Unit> {
        return try {
            // Delete related reminders first
            val relatedReminders = reminderRepository.observeByPersonId(personId).first()
            relatedReminders.forEach { reminder ->
                reminderRepository.deleteById(reminder.id)
            }

            // Delete the person
            personRepository.deleteById(personId)

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: ios/DailyMemory/Domain/UseCase/Person/DeletePersonUseCase.swift
import Foundation

/// Use case for deleting a person.
/// Also handles cleanup of related reminders.
final class DeletePersonUseCase {
    private let personRepository: PersonRepository
    private let reminderRepository: ReminderRepository

    init(
        personRepository: PersonRepository = PersonRepositoryImpl(),
        reminderRepository: ReminderRepository = ReminderRepositoryImpl()
    ) {
        self.personRepository = personRepository
        self.reminderRepository = reminderRepository
    }

    func execute(personId: String) async throws {
        // Delete related reminders first
        let relatedReminders = try await reminderRepository.getByPersonId(personId)
        for reminder in relatedReminders {
            try await reminderRepository.delete(id: reminder.id)
        }

        // Delete the person
        try await personRepository.delete(id: personId)
        await SyncManager.shared.enqueuePersonChange(id: personId, type: .delete)
    }
}
